### ISSUE TITLE
Setting default values

### ISSUE THREAD

**Issue by binary132** [xeipuuv/gojsonschema#45]
I notice that the "default" key is not tested.  Is it possible to generate a default object for a schema such as this?  If not, would you be open to a PR to generate a default map[string]interface{} from a gojsonschema document?

```json
{
  "properties": {
    "outfile": {
      "type": "string",
      "default": "out.bz2"
    },
  },
  "type": "object"
}
```

should produce

```go
map[string]interface{}{
    "outfile": "out.bz2",
}
```

as a default object.

**Comment by xeipuuv** [comment]
Oh you are right, and yes i am open to PR for sure.

**Comment by binary132** [comment]
I have added Defaults and opened a pull request against our fork here:

https://github.com/juju/gojsonschema/pull/2

Please feel free to add any comments or concerns.

Once this has merged on our fork, I will open the PR against the mainline repo since time is of the essence for integrating this feature in our fork.

**Comment by adityanatraj** [comment]
hey @binary132 I noticed that your pull request was merged into your fork, any chance you'll get around to that PR here soon?

**Comment by binary132** [comment]
Hi @adityanatraj , my involvement with that project has ceased, but I will try to find time to open a PR against this project.  We had to rush it in and I'm not sure how many changes will need to be made to fix it up for the mainline repo.  Hope that's not too inconvenient, my time is not my own at present

**Comment by tclass** [comment]
any updates on this? Would be really nice to have

**Comment by binary132** [comment]
I'll make myself a todo to get it into shape.  :)  The version we sketched up for Juju uses some abbreviated syntax to represent inline `map[string]interface{}`s and I don't think it's suitable for merge at present.  I expect I can find time for it some time in the next few weeks, in the middle of moving right now.

**Comment by sparkprime** [comment]
I'd really like this feature, and it looks like a PR was given https://github.com/xeipuuv/gojsonschema/pull/80 but was rejected.  Why not accept it, if it would close this bug?  thanks

**Comment by cristiangraz** [comment]
I usually decode json into structs, but validate to ensure the data is valid otherwise returning an error and ending the request. 

I've accomplished this really easily by just initializing the struct with default values, then decoding the json into it and then validating.

Something like this:

```go
data := SomeStruct{Outfile: "out.bz2", OtherDefault: "other-value"}
buf := new(bytes.Buffer)
tee := io.TeeReader(req.Body, buf)
if err := json.NewDecoder(tee).Decode(&data); err != nil {
	// Error unmarshaling into struct
}

body := buf.String()

document := gojsonschema.NewStringLoader(body)
result, err := gojsonschema.Validate(schema, document)
if err != nil {
	// json schema error
}
```
If there were no errors, data now has default values as well as whatever values were sent in the JSON request.

Not sure if this is helpful to anyone who comes across this, but an easy way to accomplish the same thing if you are working with structs.

**Comment by drasko** [comment]
Will this ever be merged? Looks like extremely useful enhancement! @binary132 any updates?

**Comment by binary132** [comment]
The merge from my work is not straightforward.  It's against an outdated version of `gojsonschema` and I don't have spare cycles to manage the migration.  All of the content can be seen in my branch and I'm fine with anyone else picking this up.  However, there are some dependencies on updates in other packages (gojsonpointer and gojsonreference had some problems when I was working with this.)

**Comment by codepushr** [comment]
Any updates on the open PRs?

**Comment by andy-miracl** [comment]
We have a library that support this feature, which you may find useful. The library internally uses gojsonschema. https://github.com/miracl/conflate

**Comment by andrewpmartinez** [comment]
Well, this dead trail is frustrating.

**Comment by olegsu** [comment]
So the default populate use case is not supported in this project?

**Comment by johandorland** [comment]
@olegsu No. This library supports JSON schema validation and generating a default object from a schema is something not covered in the JSON schema specification.

**Comment by robbyt** [comment]
Alternatively, I've been using https://github.com/miracl/conflate which uses a older version of gojsonschema (and only supports Draft v4). It has a few issues, but merge/defaults handling works for basic schemas.

**Comment by nick4fake** [comment]
@johandorland I understand that generating default object is not covered by the spec, but there is slight problem - we can't have another library generating it. JSON schema is very complex and it might have multiple branches, and only after one is chosen we can apply defaults to it.

If you believe it should not be a part of this library, can we at least have some generic method to provide part of the schema used for validation?

**Comment by lewisdiamond** [comment]
@xeipuuv Are you still open to PRs for this?
